lib/pbench/server/api/resources/metadata_api.py

```diff
@@ -0,0 +1,355 @@
+
+        # Verify if the metadata session id in the url belongs to the logged in user
+        self.verify_metadata(metadata_session, "get")
+
+        response_object = {
+            "message": "success",
+            "data": metadata_session.get_json(),
+        }
+        return make_response(jsonify(response_object), 200)
+
+    @Auth.token_auth.login_required()
+    def put(self, id=None):
+        """
+        Put request for updating a metadata session of a user given a metadata id.
+        This requires a Pbench auth token in the header field
+
+        The url requires a metadata session id such as /user/metadata/<int:id>
+
+        This requires a JSON data with required user metadata fields to update
+        {
+            "value": "new text value",
+            ...
+        }
+        To update the value field, it is required to send the complete text blob again
+
+        Required headers include
+            Authorization:   Bearer <Pbench authentication token (user received upon login)>
+
+        :return: JSON Payload
+            response_object = {
+                    "message": "success",
+                    "data": {
+                        "id": "metadata_id",
+                        "value": "client text blob"
+                        "created": "Datetime string"
+                        "updated": "Datetime String"
+                        "key": "key string"
+                        }
+                }
+        """
+        if not id:
+            self.logger.warning("Metadata id not provided during metadata query")
+            abort(400, message="Please provide a metadata id to query")
+
+        post_data = request.get_json()
+        if not post_data:
+            self.logger.warning("Invalid json object: {}", request.url)
+            abort(400, message="Invalid json object in request")
+
+        try:
+            metadata_session = Metadata.query(id=id)[0]
+        except Exception:
+            self.logger.exception(
+                "Exception occurred in the PUT request while querying the Metadata model, id: {}",
+                id,
+            )
+            abort(500, message="INTERNAL ERROR")
+
+        # Verify if the metadata session id in the url belongs to the logged in user
+        self.verify_metadata(metadata_session, "put")
+
+        # Check if the metadata payload contain fields that are either protected or
+        # not present in the metadata db. If any key in the payload does not match
+        # with the column name we will abort the update request.
+        non_existent = set(post_data.keys()).difference(
+            set(Metadata.__table__.columns.keys())
+        )
+        if non_existent:
+            self.logger.warning(
+                "User trying to update fields that are not present in the metadata database. Fields: {}",
+                non_existent,
+            )
+            abort(400, message="Invalid fields in update request payload")
```

Comment:
yeah we can add that

Reply:
On the second thought however, should we let the requester figure out which fields are invalid, this is in accordance with our design that we try not to help the requester debug the request, we just tell them what went wrong

Reply:
Anything around security, that's true; not so sure we want to extend that to non-security failures. I'd rather be helpful where it makes sense.

Reply:
Given that the client can issue this request over and over with different fields, it would be mechanically straightforward for the client developer to figure out what part of the request is bad...but, it would be annoying for him/her.  So, really, I see no reason why we shouldn't just indicate the offending field(s) in the message and save all the trouble.